### PR TITLE
⏺️ Recording part 1 - Add useful information to the top-bar

### PR DESCRIPTION
**Pull request by marcoambrosini**
Signed-off-by: Marco <marcoambrosini@icloud.com>

### 🖼️ Screenshots

🏚️ Before 
Screenshot 2022-11-18 at 16 20 26

🏡 After
Screenshot 2022-11-18 at 16 19 32

### 🚧 TODO

- [x] Open participants tab (not chat) when clicking on participants button
- [x] Fix conversation title color 
- [x] Fix app navigation toggle alignment
- [x] Check mobile mode, see if anything can be collapsed in the action menu
- [x] Implement the clock with a dummy unix timestamp

### 🏁 Checklist

- [ ] ⛑️ Tests (unit and/or integration) are included
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [ ] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [x] 🔖 Capability is added or not needed

### CODE REVIEW

File: src/components/TopBar/CallTime.vue [hunk 1]
```diff
@@ -0,0 +1,127 @@
+<!--
+  - @copyright Copyright (c) 2021 Marco Ambrosini <marcoambrosini@icloud.com>  -
```
Comment (SystemKeeper) [inline src/components/TopBar/CallTime.vue:2]: Time travel?

File: src/components/TopBar/TopBar.vue [hunk 2]
```diff
@@ -68,36 +75,54 @@
...
-					</template>
-				</NcActionButton>
-			</NcActions>
-		</div>
-		<NcCounterBubble v-if="!isSidebar && showOpenSidebarButton && isInCall && unreadMessagesCounter > 0"
-			class="unread-messages-counter"
-			:highlighted="hasUnreadMentions">
-			{{ unreadMessagesCounter }}
-		</NcCounterBubble>
+				:type="isInCall ? 'tertiary-on-primary': 'tertiary'"
```
Comment (nickvergessen) [inline src/components/TopBar/TopBar.vue:117]: This is wrong. the button is never on the blue/background topbar.
Then again it's inside `v-if="isInCall…"

But currently it's missing hover feedback. I think you want to use:

```suggestion
				type="tertiary"
```

File: src/components/TopBar/TopBar.vue [hunk 3]
```diff
@@ -68,36 +75,54 @@
...
+				</NcActions>
+				<NcCounterBubble v-if="!isSidebar && isInCall && unreadMessagesCounter > 0"
+					class="chat-button__unread-messages-counter"
+					:highlighted="hasUnreadMentions">
+					{{ unreadMessagesCounter }}
+				</NcCounterBubble>
+			</div>
+
+			<!-- participants button -->
+			<NcButton v-if="isInCall & !isOneToOneConversation"
```
Comment (nickvergessen) [inline src/components/TopBar/TopBar.vue:114]: I'm pretty sure you don't want to do bitflag comparisions...
```suggestion
			<NcButton v-if="isInCall && !isOneToOneConversation"
```

File: src/components/TopBar/TopBar.vue [hunk 4]
```diff
@@ -68,36 +75,54 @@
...
-		<NcCounterBubble v-if="!isSidebar && showOpenSidebarButton && isInCall && unreadMessagesCounter > 0"
-			class="unread-messages-counter"
-			:highlighted="hasUnreadMentions">
-			{{ unreadMessagesCounter }}
-		</NcCounterBubble>
+				:type="isInCall ? 'tertiary-on-primary': 'tertiary'"
+				@click="openSidebar('participants')">
+				<template #icon>
+					<AccountMultiple :size="20"
+						:fill-color="isInCall ? '#ffffff': ''" />
```
Comment (nickvergessen) [inline src/components/TopBar/TopBar.vue:121]: As said, only shows in calls:
```suggestion
						fill-color="#fffff" />
```

File: src/components/TopBar/CallTime.vue [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+  - but WITHOUT ANY WARRANTY; without even the implied warranty of
+  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  - GNU Affero General Public License for more details.
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+-->
+
+<template>
+	<NcPopover class="top-bar__button call-time"
```
Comment (nickvergessen) [inline src/components/TopBar/CallTime.vue:22]: How are we going to handle normal users?
They should only see the RecordCircle but it must not be actionable? Or we should the person that started the recording in the popover?

File: src/components/TopBar/CallTime.vue [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+		},
+	},
+
+	mounted() {
+		// Start the timer when mounted
+		setInterval(this.computeElapsedTime, 1000)
+	},
+
+	beforeDestroy() {
+		clearInterval(this.computeElapsedTime)
```
Comment (nickvergessen) [inline src/components/TopBar/CallTime.vue:141]: This is not how it works. You need to store the return of `setInterval` into data or something and inject that here on the clearInterval call. See "More examples" the second example:
https://www.w3schools.com/jsref/met_win_setinterval.asp

File: src/components/TopBar/TopBar.vue [hunk 7]
```diff
@@ -248,6 +280,10 @@ export default {
 				return !peer.sessionIds.length
 			} else return false
 		},
+
+		participantsInCall() {
+			return this.$store.getters.participantsInCall(this.token) ? this.$store.getters.participantsInCall(this.token) : ''
```
Comment (nickvergessen) [inline src/components/TopBar/TopBar.vue:285]:
```suggestion
			return this.$store.getters.participantsInCall(this.token) || ''
```
Shorter and not calling the getter twice?

File: src/components/TopBar/TopBar.vue [hunk 8]
```diff
@@ -317,14 +353,22 @@ export default {
 			}
 		},
 
-		openSidebar() {
+		openSidebar(activeTab) {
+			if (typeof activeTab === 'string') {
+				emit('spreed:select-active-sidebar-tab', activeTab)
```
Comment (nickvergessen) [inline src/components/TopBar/TopBar.vue:358]: This somehow does not set focus, right?
When I keyboard navigate to the participant icon and enter, the sidebar opens but the tab index is back on the call view?

File: src/components/TopBar/CallTime.vue [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+				{{ formattedTime }}
+			</ncbutton>
+		</template>
+		<NcButton type="tertiary-no-background"
+			:wide="true"
+			@click="stopRecording">
+			<template #icon>
+				<StopIcon :size="20" />
+			</template>
+			{{ t('spreed', 'Stop recording') }}
```
Comment (nickvergessen) [inline src/components/TopBar/CallTime.vue:48]: While the options in the NcActions are grey/max contrast/something
This one is pretty dark/black or white/bright:
Before | After
---|---
Bildschirmfoto vom 2022-11-21 11-49-19 | Bildschirmfoto vom 2022-11-21 11-49-25
Bildschirmfoto vom 2022-11-21 11-51-13 | Bildschirmfoto vom 2022-11-21 11-51-18

Also the rounding is different from the actions menu, but we can move this to a follow up, as the PR has too many things touching already.

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: src/components/TopBar/CallTime.vue (hunk 5 above)
Comment (marcoambrosini) [reply src/components/TopBar/CallTime.vue:22]: let's just hide the popover for now?

File: src/components/TopBar/TopBar.vue (hunk 7 above)
Comment (marcoambrosini) [reply src/components/TopBar/TopBar.vue:285]: that returns a boolean

Comment (nickvergessen) [reply src/components/TopBar/TopBar.vue:285]: ### Firefox
Bildschirmfoto vom 2022-11-22 11-16-19

### Chromium
Bildschirmfoto vom 2022-11-22 11-16-34

File: src/components/TopBar/CallTime.vue (hunk 5 above)
Comment (nickvergessen) [reply src/components/TopBar/CallTime.vue:22]: But we still need the trigger, right?

File: src/components/TopBar/TopBar.vue (hunk 8 above)
Comment (marcoambrosini) [reply src/components/TopBar/TopBar.vue:358]: This is the case on master too, but only with firefox. I'll create an issue

Comment (marcoambrosini) [reply src/components/TopBar/TopBar.vue:358]: #8387

File: src/components/TopBar/CallTime.vue (hunk 9 above)
Comment (nickvergessen) [reply src/components/TopBar/CallTime.vue:48]: https://github.com/nextcloud/spreed/issues/8388

File: src/components/TopBar/CallTime.vue [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+	name: 'CallTime',
+
+	components: {
+		RecordCircle,
+		StopIcon,
+		NcPopover,
+		NcButton,
+	},
+
+	mixins: [isInLobby],
```
Comment (nickvergessen) [inline src/components/TopBar/CallTime.vue:71]: Just for the record:
This `isModerator` allows guest moderators too, but they can not start the recording, as they don't have storage.

File: src/mixins/isInLobby.js [hunk 11]
```diff
@@ -25,12 +25,18 @@ import { PARTICIPANT, WEBINAR } from '../constants.js'
...
 export default {
 
 	computed: {
+		token() {
+			return this.$store.getters.getToken()
+		},
+
+		conversation() {
+			return this.$store.getters.conversation(this.token) || this.$store.getters.dummyConversation
+		},
```
Comment (nickvergessen) [inline src/mixins/isInLobby.js:38]: See feedback in chat by @ShGKme

File: src/store/participantsStore.js [hunk 12]
```diff
@@ -155,6 +155,14 @@ const getters = {
 
 		return {}
 	},
+
+	participantsInCall: (state) => (token) => {
+		if (state.inCall[token]) {
```
Comment (nickvergessen) [inline src/store/participantsStore.js:160]: As mentioned in 1-1, this is a temporary store to overwrite the "in call" flag for the current participant.
Do an array filter on `state.attendees[token]` with `attendee.inCall !== 0` and then count the length as currently.

File: src/components/TopBar/CallTime.vue (hunk 5 above)
Comment (marcoambrosini) [reply src/components/TopBar/CallTime.vue:22]: Trigger is there but disabled

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**